### ISSUE TITLE
Reset a `torch.optim.Optimizer`

### ISSUE THREAD

**Issue by aretor** [pytorch/pytorch#37410]
## 🚀 Feature
A 'reset_state' method that resets the `state` of an optimizer. 

## Motivation
Sometimes we may need to reuse an optimizer to train the same model on another dataset. Some optimizers, though, keep trace of some additional stats (e.g. Adam's moving averages), and currently, there is no method to reset these stats unless we expose some internal attributes of the class.

Check [this question](https://discuss.pytorch.org/t/reset-optimizer-stats/78516/2) at the forum for more info.

## Pitch

My proposal, consists on adding a method `reset_state(self, state=None)`. If `state is None` is `True`, then we just reset the state, i.e. `self.state = defaultdict(dict)`, otherwise `self.state = state`

## Alternatives

Either you manually retrieve all the parameters of an optimizer and create a new one or you manually reset the `state` attribute. In both cases we're exposing some internals of the optimizer.

**Comment by vadimkantorov** [comment]
(Related reset concept request but for lr: https://github.com/pytorch/pytorch/issues/32021)

**Comment by colesbury** [comment]
I'll let @vincentqb  comment

**Comment by vincentqb** [comment]
Similar functionality has also been requested in #30232. This latter solution also replaced part of the `__init__` code.

Can you provide a minimal code snippet that would represent what your ideal workflow would be? Is there a way to satisfy this issue, #32021, and the `reset_state` part of #30232?

I'm open to reviewing a pull request for this.

**Comment by Xylambda** [comment]
I'm not sure if this issue is still alive despite the `Open` label. 

I think the ideal workflow would be something like

```python
import torch.nn as nn
import torch.optim as optim


model = nn.Linear(in_features=1, out_features=1)
model.to(device)

optimizer = optim.Adam(model.parameters(),  lr=0.005)

# do stuff (...)

optimizer.reset_parameters()
```

Similar to what we have for layers.

In my case, it would be useful to easily run multiple experiments inside a `for` loop without manually accessing the internal parameters via `__dict__` or other alternatives:

```python
for seed in seeds_list:
    # set seeds
    np.random.seed(seed)
    torch.manual_seed(seed)

    # train model (train + validation)
    (...)

    # save model and optimizer states + metrics
    (...)

    # restart model and optimizer
    model.reset_parameters()
    optimizer.reset_parameters() # DISC: would we need to pass the model params again?
```

**Comment by vadimkantorov** [comment]
Given that optimizers are becoming more functional now, I think it may make sense to support the same interface for them. It may even make sense to make optimizers regular modules?

**Comment by Xylambda** [comment]
I don't feel qualified to assess if the optimizers should be modules because I don't know the source code in detail. To me, it seems a good (and cool!) idea but also I feel it should be difficult to rearrange the API in order to support the new design. 

I would like to read the opinions of more educated developers than me, though. With pros and cons.

Anyway, I believe the necessity of a `reset_parameters` or `reset_state` in the optimizer is real and PyTorch should support it.

**Comment by boykovdn** [comment]
Hi, I've ran into a use case for resetting the state of an optimizer (Adam in my case), for when I want to freeze a part of my network. It seems the momentum carries on updating the layer parameters even after I freeze them (set requires_grad = False), and it would be nice to call reset() to remove the accumulated momentum off the optimizer without changing internal state. I am using pytorch 1.9.1 and setting the learning rate to 0 in Adam's param_groups to stop the parameters updating (I have a separate optimizer for this part of the network).

**Comment by carlosgmartin** [comment]
@aretor A simple `reset` method sounds great. Any reason for the longer identifier?

**Comment by jpgard** [comment]
Just wanted to give a +1 for a reset method.

For one more example of an application, you can see [this line of code](https://github.com/facebookresearch/DomainBed/blob/main/domainbed/algorithms.py#L351) of the implementation of Invariant Risk Minimization in DomainBed. 

Here, there is a need to reset Adam after a warm-starting period due to a sharp jump in the gradients. While this library simply reassigns the optimizer to a class attribute (`self.optimizer`), this isn't always how models are implemented -- it's common to decouple the optimizer from the model object, in which case an `optimizer.reset()` method would be very useful.

**Comment by nalzok** [comment]
Additionally, when using frameworks like Lightning Fabric, you have to register the optimizer to a `Fabric` object by doing `model, optimizer = fabric.setup(model, optimizer)`. If we simply (re)assign `optimizer` to a new optimizer, the new one would not be registered correctly.

**Comment by ryanliu30** [comment]
A `reset_state` will also be very helpful to my use case. I am using `pytorch_lighting` to train my model where I will have to periodically reset some of my submodules' weights and the corresponding optimizer's weights. If I just create a new optimizer it won't be registered to `pytorch_lighting` and could possibly cause some error.
